.github/workflows/pypi-package.yml

```diff
@@ -0,0 +1,72 @@
+---
+name: Build and upload package
+
+on:
+  push:
+    branches: [main]
+    tags: ["*"]
+  release:
+    types:
+      - published
+  workflow_dispatch:
+
+permissions:
+  contents: read
+  id-token: write
+
+jobs:
+  build-package:
+    name: Build & check package
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        submodules: recursive
+        fetch-depth: 0
+    - name: Set up Python 3.12
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.12
+    - run: pip install --upgrade build twine
+    - name: Build sdist and wheel
+      run: python -m build
+    - name: Check package
+      run: twine check dist/*
+    - name: Upload sdist and wheel artifacts
+      uses: actions/upload-artifact@v4
+      with:
+        name: dist
+        path: dist/
+
+  auto-release-test-pypi:
+    runs-on: ubuntu-latest
+    environment: "Test-release"
+    needs: [build-package]
+    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          name: dist
+          path: dist/
+      - uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+          user: __token__
+          password: ${{ secrets.TESTPYPI_API_TOKEN }}
```

Comment:
@effigies @oesteban Do we also have an authentication token for TestPyPI? (For now, I've used this generic variable name.)